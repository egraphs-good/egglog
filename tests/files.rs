use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
