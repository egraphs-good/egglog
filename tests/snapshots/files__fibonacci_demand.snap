---
source: tests/files.rs
expression: serialized
---
nodes:
  i64-4208978898528913939:
    op: "7"
    children: []
    eclass: i64-4208978898528913939
    cost: 0
  Fib-4208978898528913939:
    op: Fib
    children:
      - i64-4208978898528913939
    eclass: "0"
    cost: 1
  i64-16783941965674463102:
    op: "6"
    children: []
    eclass: i64-16783941965674463102
    cost: 0
  Fib-16783941965674463102:
    op: Fib
    children:
      - i64-16783941965674463102
    eclass: "1"
    cost: 1
  i64-10912160959110460649:
    op: "5"
    children: []
    eclass: i64-10912160959110460649
    cost: 0
  Fib-10912160959110460649:
    op: Fib
    children:
      - i64-10912160959110460649
    eclass: "2"
    cost: 1
  i64-5040379952546458196:
    op: "4"
    children: []
    eclass: i64-5040379952546458196
    cost: 0
  Fib-5040379952546458196:
    op: Fib
    children:
      - i64-5040379952546458196
    eclass: "4"
    cost: 1
  i64-17615343019692007359:
    op: "3"
    children: []
    eclass: i64-17615343019692007359
    cost: 0
  Fib-17615343019692007359:
    op: Fib
    children:
      - i64-17615343019692007359
    eclass: "6"
    cost: 1
  i64-11743562013128004906:
    op: "2"
    children: []
    eclass: i64-11743562013128004906
    cost: 0
  Fib-11743562013128004906:
    op: Fib
    children:
      - i64-11743562013128004906
    eclass: "8"
    cost: 1
  i64-0:
    op: "0"
    children: []
    eclass: i64-0
    cost: 0
  Fib-0:
    op: Fib
    children:
      - i64-0
    eclass: "12"
    cost: 1
  i64-5871781006564002453:
    op: "1"
    children: []
    eclass: i64-5871781006564002453
    cost: 0
  Fib-5871781006564002453:
    op: Fib
    children:
      - i64-5871781006564002453
    eclass: "8"
    cost: 1
  Num-0:
    op: Num
    children:
      - i64-0
    eclass: "12"
    cost: 1
  Num-5871781006564002453:
    op: Num
    children:
      - i64-5871781006564002453
    eclass: "8"
    cost: 1
  Num-11743562013128004906:
    op: Num
    children:
      - i64-11743562013128004906
    eclass: "6"
    cost: 1
  Num-17615343019692007359:
    op: Num
    children:
      - i64-17615343019692007359
    eclass: "4"
    cost: 1
  Num-10912160959110460649:
    op: Num
    children:
      - i64-10912160959110460649
    eclass: "2"
    cost: 1
  i64-10080759905092916392:
    op: "8"
    children: []
    eclass: i64-10080759905092916392
    cost: 0
  Num-10080759905092916392:
    op: Num
    children:
      - i64-10080759905092916392
    eclass: "1"
    cost: 1
  i64-2546176790493825425:
    op: "13"
    children: []
    eclass: i64-2546176790493825425
    cost: 0
  Num-2546176790493825425:
    op: Num
    children:
      - i64-2546176790493825425
    eclass: "0"
    cost: 1
  f7-0:
    op: f7
    children: []
    eclass: "0"
    cost: 9223372036854776000
  Add-7659469028595837896:
    op: Add
    children:
      - Fib-16783941965674463102
      - Fib-10912160959110460649
    eclass: "0"
    cost: 1
  Add-15318938057191675792:
    op: Add
    children:
      - Num-10912160959110460649
      - Fib-5040379952546458196
    eclass: "1"
    cost: 1
  Add-17179538364779065510:
    op: Add
    children:
      - Num-17615343019692007359
      - Fib-17615343019692007359
    eclass: "2"
    cost: 1
  Add-16545935510313822457:
    op: Add
    children:
      - Num-11743562013128004906
      - Fib-11743562013128004906
    eclass: "4"
    cost: 1
  Add-3337369588703030241:
    op: Add
    children:
      - Num-5871781006564002453
      - Num-0
    eclass: "8"
    cost: 1
  Add-16743733709866123661:
    op: Add
    children:
      - Add-3337369588703030241
      - Fib-11743562013128004906
    eclass: "6"
    cost: 1
root_eclasses: []
class_data:
  "0":
    type: Expr
  i64-4208978898528913939:
    type: i64
  "1":
    type: Expr
  i64-16783941965674463102:
    type: i64
  "2":
    type: Expr
  i64-10912160959110460649:
    type: i64
  "4":
    type: Expr
  i64-5040379952546458196:
    type: i64
  "6":
    type: Expr
  i64-17615343019692007359:
    type: i64
  "8":
    type: Expr
  i64-11743562013128004906:
    type: i64
  "12":
    type: Expr
  i64-0:
    type: i64
  i64-5871781006564002453:
    type: i64
  i64-10080759905092916392:
    type: i64
  i64-2546176790493825425:
    type: i64

